File: cli/crypto.cpp
#include "crypto.h"

#include <sodium.h>

#include <stdexcept>
#include <string>

namespace blin {

namespace {

constexpr std::size_t kEnvNonceBytes = crypto_aead_xchacha20poly1305_ietf_NPUBBYTES;
constexpr std::size_t kEnvMacBytes   = crypto_aead_xchacha20poly1305_ietf_ABYTES;
constexpr std::uint8_t kEnvVersion   = 1;

constexpr std::size_t kChalPlainBytes  = 33;
constexpr std::size_t kChalCipherBytes = 49;
constexpr std::size_t kChalBurnOffset  = 32;

constexpr char kSeedAlphabet[] =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

}  // namespace

void cryptoInit() {
    if (sodium_init() < 0) {
        throw std::runtime_error("sodium_init failed");
    }
}

Bytes deriveKey(const std::string& seed, const std::string& password) {
    const std::string saltInput = "cryptoblin/v1/" + seed;
    Bytes salt(crypto_pwhash_SALTBYTES);
    crypto_generichash(salt.data(), salt.size(),
                       reinterpret_cast<const std::uint8_t*>(saltInput.data()),
                       saltInput.size(), nullptr, 0);

    const std::string pwd = seed + ":" + password;
    Bytes out(32);
    if (crypto_pwhash(out.data(), out.size(),
                      pwd.c_str(), pwd.size(),
                      salt.data(),
                      crypto_pwhash_OPSLIMIT_INTERACTIVE,
                      crypto_pwhash_MEMLIMIT_INTERACTIVE,
                      crypto_pwhash_ALG_ARGON2ID13) != 0) {
        throw std::runtime_error("deriveKey: out of memory");
    }
    return out;
}

Challenge makeChallenge(const Bytes& key, bool burn) {
    Challenge c;
    c.plaintext.assign(kChalPlainBytes, 0);
    randombytes_buf(c.plaintext.data(), kChalBurnOffset);
    c.plaintext[kChalBurnOffset] = burn ? 1 : 0;

    c.nonce.assign(kEnvNonceBytes, 0);
    randombytes_buf(c.nonce.data(), c.nonce.size());

    c.ciphertext.assign(kChalCipherBytes, 0);
    unsigned long long ctLen = 0;
    if (crypto_aead_xchacha20poly1305_ietf_encrypt(
            c.ciphertext.data(), &ctLen,
            c.plaintext.data(), c.plaintext.size(),
            nullptr, 0, nullptr,
            c.nonce.data(), key.data()) != 0) {
        throw std::runtime_error("makeChallenge: encrypt failed");
    }
    c.ciphertext.resize(ctLen);
    return c;
}

Bytes solveChallenge(const Bytes& nonce, const Bytes& ciphertext, const Bytes& key) {
    if (nonce.size() != kEnvNonceBytes || ciphertext.size() < kEnvMacBytes) {
        throw std::runtime_error("solveChallenge: bad sizes");
    }
    Bytes out(ciphertext.size());
    unsigned long long ptLen = 0;
    if (crypto_aead_xchacha20poly1305_ietf_decrypt(
            out.data(), &ptLen, nullptr,
            ciphertext.data(), ciphertext.size(),
            nullptr, 0,
            nonce.data(), key.data()) != 0) {
        throw std::runtime_error("solveChallenge: decrypt failed (wrong key/password?)");
    }
    out.resize(ptLen);
    return out;
}

Bytes sealEnvelope(const Bytes& plaintext, const Bytes& key) {
    Bytes nonce(kEnvNonceBytes);
    randombytes_buf(nonce.data(), nonce.size());

    Bytes ct(plaintext.size() + kEnvMacBytes);
    unsigned long long ctLen = 0;
    if (crypto_aead_xchacha20poly1305_ietf_encrypt(
            ct.data(), &ctLen,
            plaintext.data(), plaintext.size(),
            nullptr, 0, nullptr,
            nonce.data(), key.data()) != 0) {
        throw std::runtime_error("sealEnvelope: encrypt failed");
    }
    ct.resize(ctLen);

    Bytes out(1 + kEnvNonceBytes + ct.size());
    out[0] = kEnvVersion;
    std::copy(nonce.begin(), nonce.end(), out.begin() + 1);
    std::copy(ct.begin(), ct.end(), out.begin() + 1 + kEnvNonceBytes);
    return out;
}

Bytes openEnvelope(const Bytes& env, const Bytes& key) {
    if (env.size() < 1 + kEnvNonceBytes + kEnvMacBytes) {
        throw std::runtime_error("openEnvelope: too short");
    }
    if (env[0] != kEnvVersion) {
        throw std::runtime_error("openEnvelope: unknown version");
    }
    const std::uint8_t* nonce = env.data() + 1;
    const std::uint8_t* ct    = env.data() + 1 + kEnvNonceBytes;
    const std::size_t   ctLen = env.size() - 1 - kEnvNonceBytes;

    Bytes pt(ctLen);
    unsigned long long ptLen = 0;
    if (crypto_aead_xchacha20poly1305_ietf_decrypt(
            pt.data(), &ptLen, nullptr,
            ct, ctLen, nullptr, 0,
            nonce, key.data()) != 0) {
        throw std::runtime_error("openEnvelope: decrypt failed");
    }
    pt.resize(ptLen);
    return pt;
}

std::string randomSeed() {
    std::string s(8, '0');
    Bytes r(8);
    randombytes_buf(r.data(), r.size());
    for (std::size_t i = 0; i < 8; ++i) {
        s[i] = kSeedAlphabet[r[i] % 62];
    }
    return s;
}

}  // namespace blin

File: web/src/lib/crypto.js
// All crypto runs locally via libsodium-wrappers-sumo (WASM), so the site
// works correctly over plain http://. The server never sees the key, the
// seed, or the password.

import _sodium from 'libsodium-wrappers-sumo';

let sodium = null;
let readyPromise = null;

export function ready() {
    if (!readyPromise) {
        readyPromise = (async () => {
            await _sodium.ready;
            sodium = _sodium;
        })();
    }
    return readyPromise;
}

const KEY_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

export function generateKeySeed() {
    const bytes = new Uint8Array(8);
    crypto.getRandomValues(bytes);
    let out = '';
    for (let i = 0; i < 8; ++i) out += KEY_ALPHABET[bytes[i] % 62];
    return out;
}

const ENCODER = new TextEncoder();

export async function deriveKey(seed, password = '') {
    await ready();
    const salt = sodium.crypto_generichash(
        sodium.crypto_pwhash_SALTBYTES,
        ENCODER.encode('cryptoblin/v1/' + seed)
    );
    const pwd = ENCODER.encode(seed + ':' + password);
    return sodium.crypto_pwhash(
        32, pwd, salt,
        sodium.crypto_pwhash_OPSLIMIT_INTERACTIVE,
        sodium.crypto_pwhash_MEMLIMIT_INTERACTIVE,
        sodium.crypto_pwhash_ALG_ARGON2ID13
    );
}

// Proof-of-key challenge. Plaintext = 32 random + 1 burn-flag byte; the
// burn bit is AEAD-bound so the decrypter sees the value the creator set.
export const CHALLENGE_PLAIN_BYTES = 33;
export const CHALLENGE_BURN_OFFSET = 32;

export async function makeChallenge(key, { burn = false } = {}) {
    await ready();
    const plaintext = new Uint8Array(CHALLENGE_PLAIN_BYTES);
    plaintext.set(sodium.randombytes_buf(CHALLENGE_BURN_OFFSET), 0);
    plaintext[CHALLENGE_BURN_OFFSET] = burn ? 1 : 0;
    const nonce     = sodium.randombytes_buf(
        sodium.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES
    );
    const ciphertext = sodium.crypto_aead_xchacha20poly1305_ietf_encrypt(
        plaintext, null, null, nonce, key
    );
    return { plaintext, nonce, ciphertext };
}

// Decrypt a challenge served by /open. Throws on auth failure (wrong key,
// wrong password, or a decoy from the server). The caller cannot tell those
// three cases apart — that is the whole point.
export async function solveChallenge(nonceBytes, ciphertextBytes, key) {
    await ready();
    return sodium.crypto_aead_xchacha20poly1305_ietf_decrypt(
        null, ciphertextBytes, null, nonceBytes, key
    );
}

// Seal the (binary) payload as the encrypted blob. Output layout:
//   [VERSION byte] [NONCE bytes] [ciphertext + MAC]
// — that is what gets POSTed (after the challenge prefix) to /api/pastes.

const ENVELOPE_VERSION        = 1;        // wire-format identifier
const ENVELOPE_VERSION_BYTES  = 1;        // byte size of the version field
const ENVELOPE_NONCE_BYTES    = 24;       // crypto_aead_xchacha20poly1305_ietf_NPUBBYTES
const ENVELOPE_MAC_BYTES      = 16;       // crypto_aead_xchacha20poly1305_ietf_ABYTES
export const ENVELOPE_OVERHEAD_BYTES =
    ENVELOPE_VERSION_BYTES + ENVELOPE_NONCE_BYTES + ENVELOPE_MAC_BYTES;

const NONCE_OFFSET = ENVELOPE_VERSION_BYTES;
const CIPHER_OFFSET = ENVELOPE_VERSION_BYTES + ENVELOPE_NONCE_BYTES;

export async function sealEnvelope(plaintextBytes, key) {
    await ready();
    const nonce = sodium.randombytes_buf(ENVELOPE_NONCE_BYTES);
    const ct = sodium.crypto_aead_xchacha20poly1305_ietf_encrypt(
        plaintextBytes, null, null, nonce, key
    );
    const out = new Uint8Array(ENVELOPE_VERSION_BYTES + nonce.length + ct.length);
    out[0] = ENVELOPE_VERSION;
    out.set(nonce, NONCE_OFFSET);
    out.set(ct, CIPHER_OFFSET);
    return out;
}

export async function openEnvelope(envelopeBytes, key) {
    await ready();
    if (!envelopeBytes || envelopeBytes.length < ENVELOPE_OVERHEAD_BYTES) {
        throw new Error('envelope_too_short');
    }
    if (envelopeBytes[0] !== ENVELOPE_VERSION) throw new Error('envelope_version');
    const nonce = envelopeBytes.subarray(NONCE_OFFSET, CIPHER_OFFSET);
    const ct    = envelopeBytes.subarray(CIPHER_OFFSET);
    return sodium.crypto_aead_xchacha20poly1305_ietf_decrypt(
        null, ct, null, nonce, key
    );
}

export async function toB64(bytes) {
    await ready();
    return sodium.to_base64(bytes, sodium.base64_variants.ORIGINAL);
}

export async function fromB64(s) {
    await ready();
    return sodium.from_base64(s, sodium.base64_variants.ORIGINAL);
}
